src/commands/decompress.rs

```diff
@@ -284,6 +290,49 @@ pub fn decompress_file(options: DecompressOptions) -> crate::Result<()> {
     Ok(())
 }
 
+fn execute_decompression(
+    unpack_fn: impl FnOnce(&Path) -> crate::Result<usize>,
+    output_dir: &Path,
+    output_file_path: &Path,
+    question_policy: QuestionPolicy,
+    is_output_dir_explicit: bool,
+) -> crate::Result<ControlFlow<(), usize>> {
+    if is_output_dir_explicit {
+        unpack(unpack_fn, output_dir, question_policy)
+    } else {
+        smart_unpack(unpack_fn, output_dir, output_file_path, question_policy)
+    }
+}
+
+/// Unpacks an archive creating the output directory, this function will create the output_dir
+/// directory or replace it if it already exists. The `output_dir` needs to be empty
+/// - If `output_dir` does not exist OR is a empty directory, it will unpack there
+/// - If `output_dir` exist OR is a directory not empty, the user will be asked what to do
+fn unpack(
+    unpack_fn: impl FnOnce(&Path) -> crate::Result<usize>,
+    output_dir: &Path,
+    question_policy: QuestionPolicy,
+) -> crate::Result<ControlFlow<(), usize>> {
+    let is_valid_output_dir = !output_dir.exists() || (output_dir.is_dir() && output_dir.read_dir()?.count() == 0);
```

Comment:
`.read_dir()?.count()` reads all elements in a directory.

Can you check `.count()` calls in the project and optimize them.

Should be `.next().is_none()`, or, when it's not compared to `0`, `.nth(N).is_some/none()`